### PR TITLE
TGUI Parity with Upstream and Remaining React Porting & Polish

### PR DESCRIPTION
**Pull request by Drulikar**
# About the pull request

This PR is a follow up to #5493 and ultimately is aimed at bringing us back into parity with TG's version of the TGUI components again. For the most part this PR is done via diffing TG's TGUI folder with ours. The PRs listed below are what I found in blames and I double checked to see what files if any they affected outside the TGUI folder.

**It is recommended to close VSC and run bin/clean.cmd and then bin/build.cmd if you try out this branch**. You will be using the wrong node & prettier version otherwise.

This at least partially ports these PRs (and likely many others - but these I explicitly checked for how they affected other files):
- https://github.com/tgstation/tgstation/pull/67935
- https://github.com/tgstation/tgstation/pull/74638
- https://github.com/tgstation/tgstation/pull/75029
- https://github.com/tgstation/tgstation/pull/75431
- https://github.com/tgstation/tgstation/pull/75463
- https://github.com/tgstation/tgstation/pull/75534
- https://github.com/tgstation/tgstation/pull/78879
- https://github.com/tgstation/tgstation/pull/79322
- https://github.com/tgstation/tgstation/pull/79895
- https://github.com/tgstation/tgstation/pull/79898
- https://github.com/tgstation/tgstation/pull/79974
- https://github.com/tgstation/tgstation/pull/79991
- https://github.com/tgstation/tgstation/pull/80044
- https://github.com/tgstation/tgstation/pull/80057
- https://github.com/tgstation/tgstation/pull/80189
- https://github.com/tgstation/tgstation/pull/80304
- https://github.com/tgstation/tgstation/pull/80430
- https://github.com/tgstation/tgstation/pull/80453
- https://github.com/tgstation/tgstation/pull/80454
- https://github.com/tgstation/tgstation/pull/80920
- https://github.com/tgstation/tgstation/pull/81354
- https://github.com/tgstation/tgstation/pull/81381
- https://github.com/tgstation/tgstation/pull/81495
- https://github.com/tgstation/tgstation/pull/82334
- https://github.com/tgstation/tgstation/pull/82377
- https://github.com/tgstation/tgstation/pull/82385
- https://github.com/tgstation/tgstation/pull/82417
- https://github.com/tgstation/tgstation/pull/82420
- https://github.com/tgstation/tgstation/pull/82527
- https://github.com/tgstation/tgstation/pull/82571
- https://github.com/tgstation/tgstation/pull/82792
- https://github.com/tgstation/tgstation/pull/82854
- https://github.com/tgstation/tgstation/pull/83098
- https://github.com/tgstation/tgstation/pull/83096
- https://github.com/tgstation/tgstation/pull/83218

Additionally it ports the new base components from: 
- https://github.com/tgstation/tgstation/pull/71531
- https://github.com/tgstation/tgstation/pull/71847
- https://github.com/tgstation/tgstation/pull/74544
- https://github.com/tgstation/tgstation/pull/74704 (just a fix to above)
- https://github.com/tgstation/tgstation/pull/80743
- https://github.com/tgstation/tgstation/pull/82181 (just a fix to above)
- https://github.com/tgstation/tgstation/pull/82533
- https://github.com/tgstation/tgstation/pull/83179 (just a fix to above)

But these are missing documentation in the components readme, and we do not have any interface that utilizes them yet. I wanted to explicitly mention them separately so people can see what they can do.

Changes I still intend to do in this PR or investigate:
- [X] Fix Button.Confirm not properly handling clicks (Sort of stumped on this atm)
- [X] Fix Relay Viewer buttons not always updating
- [X] Port TGSay 1.1 & TextArea changes (https://github.com/tgstation/tgstation/pull/75431)
- [X] Fix ByondUI in Drop pod panel
- [X] Check parity with TG again and port anything else that may have been done during the creation of this PR
- [X] Find a way to prevent TGSay passthrough when hotkeying InputList (and maybe others)? https://github.com/tgstation/tgstation/pull/75463

Other notable changes not directly related to the above ports, but done as I was fixing other problems:
- Strip panel is now larger with sprites at 64x64 instead of 48 because as an Image component now they didn't alias the same.
- VOX panel's second tab now doesn't overlap the other buttons
- Wiring panel from https://github.com/cmss13-devs/cmss13/pull/1623 is now restored to use circular indicators again instead of the regression to ColorBox in https://github.com/cmss13-devs/cmss13/pull/2641
- Smart fridge interface is more in line with sorted inventories as far as layout & tooltips
- Tacmap drawing interface is tweaked a little to deal with the changes to dropdowns + clicking the tab is now disabled again until the tacmap is ready (It needs minimap subsystem to fire to have icons when flattening the map to be up to date)
- Fixes runtimes when quickly switching between tgsay categories
- Basically reverts https://github.com/cmss13-devs/cmss13/pull/5651 in favor of how TG ended up solving it via https://github.com/tgstation/tgstation/pull/81495
- Rolls back SWC to the version TG is currently using because of https://github.com/cmss13-devs/cmss13/pull/6196#discussion_r1581825437
- Adds these commands from https://github.com/tgstation/tgstation/pull/81381 though I also have it check jsx files that somereason TG doesn't do (needs to have been built once to function - just run build.bat): 
image
- Chat: Adds a clear chat button clearing the current tab, mute button to disable unread notifications for a tab, and updates the word highlight regex for case sensitivity. I left out one of the Cyrillic fixes since we shouldn't normally be handling Cyrillic  anyways.
- Ship manipulator window can now replace any existing shuttle and generally is polished up more than it was.
- Ports TGSay 1.1 though I discarded basically all the layout changes they did.
- TGUI modals can now accept a ui_state in case you want an alert for say a mob that might qdel while it is open.
- Finishes porting the reservation changes to shuttles.dm from https://github.com/tgstation/tgstation/pull/73261 that some reason were absent from https://github.com/cmss13-devs/cmss13/pull/5165
- Fixes F5 refreshing a window if default is prevented (e.g. when the hotkey interface is waiting for a key press).
- All depreciated useLocalState is now removed. In almost all situations I could useState, but a few things for MFD on CAS I needed to use sharedState.

# Explain why it's good for the game
Prepares us for 516 (though still more changes will be necessary) and finalizes the react port.

# Testing Photographs and Procedure
<details>
<summary>Screenshots & Videos</summary>

New list modal w/ duplicate key handling:
image

Fixed VOX panel (sound files used to overlap type select):
image

Updated shuttle manipulator window:
image
image
image

With replace any shuttle support: 
image

Erase current tab chat history:
image

Working case sensitivity highlights: 
image

Unread counter toggle per tab:
image

Larger strip panel:
image

Tacmap panel polish (button positioning + dropdown text color when highlighted):
image

Minor tweak to camera panel:
image

Dropdowns will now display above if too little space below: 
image

Kitchen sink works again (F12 w/ tgui-dev):
image

Layout mode works again (F11 w/ tgui-dev):
image

New checkbox modal (with theme support):
image
checkbox

New handling for button sizes in alert modals (stacked list since options wanted to use more horizontal space than limit):
image

Polished smart fridge + restore wiring panel:
smart fridge

Fixed byondui in supply pod menu:
image

TG Say:
image

Some unchanged panels (to just show working):
image
image
image
image
image
image
image
image
image
image
image
image
image
image
image
image
image
image
image
image
image
image
image
dropship

</details>

# Changelog
:cl: Drathek, mullenpaul
ui: Finalizes the TGUI React port and brings us back into parity again with TG's interface components
ui: Strip panel is now larger with sprites at 64x64 instead of 48 because as an Image component now they didn't alias the same.
ui: VOX panel's second tab now doesn't overlap the other buttons
ui: Wiring panel for some machines is now restored to use circular indicators again
ui: Smart fridge interface is more in line with sorted inventories as far as layout & tooltips
ui: Tacmap drawing interface is tweaked a little to deal with the changes to dropdowns + clicking the tab is now disabled again until the tacmap is ready (It needs minimap subsystem to fire to have icons when flattening the map to be up to date)
ui: Reworks RestrictedInput number handling
ui: Adds the DMIcon, VirtualList (deferred lazy list), Dialogue, MenuBar, StyleableSection, and Checkbox TGUI components (but nothing uses them yet)
ui: Chat: Adds a clear chat button clearing the current tab, mute button to disable unread notifications for a tab, and updates the word highlight regex for case sensitivity
ui: Updates the ship manipulator window to allow replacing any Shuttle
ui: Fixes the supply pod panel not displaying its byondui map correctly
ui: Fixes the research terminal duplicating papers in display old mode.
ui: Updates the tgui_input_list to for new features like hotkey selection, duplicate key handling, etc
fix: Fixes thinking runtimes when quickly switching between tgsay categories
code: Shuttle subsystem now is less aggressive with reservations (skipping deletion if below a threshold, delaying reservation if above a threshold)
fix: Fixes F5 key still refreshing a TGUI window even if default was set to be prevented (e.g. Hotkey interface)
/:cl:

### CODE REVIEW

**Comment by Drulikar** [comment]
Testing Photographs and Procedure has been updated with many examples

**Comment by cm13-github** [comment]
This pull request has conflicts, please resolve those before we can evaluate the pull request.

**Comment by cm13-github** [comment]
Conflicts have been resolved. A maintainer will review the pull request shortly.

File: tgui/packages/tgui/interfaces/ShuttleManipulator.jsx
Comment (mullenpaul) [inline tgui/packages/tgui/interfaces/ShuttleManipulator.jsx:1]: Why is this being migrated from tsx to jsx?

File: tgui/packages/tgui/interfaces/ShuttleManipulator.jsx [hunk 1]
```diff
@@ -144,80 +116,87 @@ const ShuttleManipulatorStatus = (props) => {
...
   );
+  const actualTemplates = templateObject[selectedTemplateId]?.templates || [];
+
   return (
     <Section>
       <Flex>
         <Flex.Item>
           <Tabs vertical>
-            {templateObject.map((template) => (
+            {map(templateObject, (template, templateId) => (
```
Comment (mullenpaul) [inline tgui/packages/tgui/interfaces/ShuttleManipulator.jsx:133]: This isn't neccesary

**Review by mullenpaul — CHANGES_REQUESTED** [review]
I've gotten to the end of the tgui/interfaces folder. Githubs UI is horrific for trying to deal with this.

File: code/controllers/subsystem/shuttles.dm
[the remark this answers is not in the document]
Comment (Drulikar) [reply code/controllers/subsystem/shuttles.dm:1]: This file is the backend for ShuttleManipulator.jsx which I had ported over.

File: code/modules/droppod/droppod_ui.dm
[the remark this answers is not in the document]
Comment (Drulikar) [reply code/modules/droppod/droppod_ui.dm:1]: I don't know about easier, but when I looked I did not find any instanced thing that appeared appropriate to add the component to. There is no structure associated that the component can just sit on. I'd need to have it I guess added to the datum each time it is opened, and removed (if not auto removed) when closed? But I'm unsure exactly what is shared between users ever this terminal is used by two people. I just figured it was easier to fix the missing logic.

File: code/modules/tgui/tgui_input_list.dm [hunk 2]
```diff
@@ -21,25 +21,34 @@
...
+ * * default - If an option is already preselected on the UI. Current values, etc.
+ * * ui_state - The TGUI UI state that will be returned in ui_state(). Default: always_state
  */
-/proc/tgui_input_list(mob/user, message, title, list/buttons, timeout = 0, theme = null)
+/proc/tgui_input_list(mob/user, message, title = "Select", list/buttons, timeout = 0, theme = null, default, ui_state = GLOB.always_state)
 	if (!user)
 		user = usr
 	if(!length(buttons))
-		return
+		return null
```
[the remark this answers is not in the document]
Comment (Drulikar) [reply code/modules/tgui/tgui_input_list.dm:37]: Sure if you are fine with stack_trace then so you can even see what call was made to rectify it.

File: tgui/packages/common/collections.ts
[the remark this answers is not in the document]
Comment (Drulikar) [reply tgui/packages/common/collections.ts:1]: CameraConsole.tsx, helpers.ts (Orbit), and index.tsx(Orbit) currently use the filter here.

vector.ts (part of https://github.com/tgstation/tgstation/pull/82417 - which is also the same PR that is modifying this same file), selectors.ts (chat), Chart.tsx, CardMod.jsx, Filteriffic.jsx, HiveLeaders.jsx, ShuttleManipulator.jsx, WeaponStats.jsx currently use the map here.

File: tgui/packages/tgui-panel/chat/ChatPanel.jsx [hunk 3]
```diff
@@ -46,7 +47,7 @@ export class ChatPanel extends Component {
       !prevProps || shallowDiffers(this.props, prevProps);
     if (shouldUpdateStyle) {
       chatRenderer.assignStyle({
-        'width': '100%',
+        width: '100%',
```
[the remark this answers is not in the document]
Comment (Drulikar) [reply tgui/packages/tgui-panel/chat/ChatPanel.jsx:50]: These are automatically removed. You can put the quotes back, then save the file, and they will be automatically removed.

File: tgui/packages/tgui-say/TguiSay.tsx
[the remark this answers is not in the document]
Comment (Drulikar) [reply tgui/packages/tgui-say/TguiSay.tsx:1]: See https://github.com/tgstation/tgstation/pull/75431/files

File: tgui/packages/tgui/interfaces/ShuttleManipulator.jsx
Comment (Drulikar) [reply tgui/packages/tgui/interfaces/ShuttleManipulator.jsx:1]: On upstream it is still a jsx https://github.com/tgstation/tgstation/commits/master/tgui/packages/tgui/interfaces/ShuttleManipulator.jsx which is what I ported.

File: tgui/packages/tgui/interfaces/Autodispenser.jsx [hunk 4]
```diff
@@ -159,10 +174,10 @@ export const Autodispenser = () => {
                 <NoticeBox>
                   {status === 2
                     ? 'Output: Internal buffer:' +
```
[the remark this answers is not in the document]
Comment (Drulikar) [reply tgui/packages/tgui/interfaces/Autodispenser.jsx:176]: This change would not comply with https://eslint.org/docs/latest/rules/no-template-curly-in-string

File: tgui/packages/tgui/interfaces/ShuttleManipulator.jsx (hunk 1 above)
Comment (Drulikar) [reply tgui/packages/tgui/interfaces/ShuttleManipulator.jsx:133]: Reversion results in 
image
and ultimately its just what upstream has here: https://github.com/tgstation/tgstation/blob/master/tgui/packages/tgui/interfaces/ShuttleManipulator.jsx#L107

File: tgui/packages/tgui-say/components/dragzone.tsx
[the remark this answers is not in the document]
Comment (Drulikar) [reply tgui/packages/tgui-say/components/dragzone.tsx:1]: See https://github.com/tgstation/tgstation/pull/75431/files

File: code/controllers/subsystem/shuttles.dm
[the remark this answers is not in the document]
Comment (Drulikar) [reply code/controllers/subsystem/shuttles.dm:1]: Of note https://github.com/cmss13-devs/cmss13/pull/5165 is what should have had these changes when it was porting https://github.com/tgstation/tgstation/commit/f88edef0fb0c6988de0381d7c29e967ce04310dd#diff-2c368e7c924b84206c882acaf418dad7ebf3c7d8f14e7845b3eef674fc82d276 (it had the mapping.dm changes, but now I have the remaining shuttle.dm changes)

**Comment by cm13-github** [comment]
This pull request has conflicts, please resolve those before we can evaluate the pull request.

**Comment by cm13-github** [comment]
Conflicts have been resolved. A maintainer will review the pull request shortly.

File: tgui/packages/tgui/interfaces/ShuttleManipulator.jsx
Comment (Drulikar) [reply tgui/packages/tgui/interfaces/ShuttleManipulator.jsx:1]: It is now tsxified again; I ought to port it upstream at some point.

**Review by harryob — APPROVED** [review]

**Pull request merged**